src/front/glsl/parser/types.rs

```diff
@@ -10,45 +10,76 @@ use crate::{
-            if let Some(Token { meta: end_meta, .. }) =
-                self.bump_if(parser, TokenValue::RightBracket)
-            {
-                meta.subsume(end_meta);
-                return Ok(Some((ArraySize::Dynamic, meta)));
-            }
+        span: &mut Span,
+        ty: &mut Handle<Type>,
+    ) -> Result<()> {
+        while self.parse_array_specifier_single(parser, span, ty)? {}
+        Ok(())
+    }
 
-            let (value, span) = self.parse_uint_constant(parser)?;
-            let constant = parser.module.constants.fetch_or_append(
-                crate::Constant {
+    /// Implementation of [`Self::parse_array_specifier`] for a single array_specifier
+    fn parse_array_specifier_single(
+        &mut self,
+        parser: &mut Parser,
+        span: &mut Span,
+        ty: &mut Handle<Type>,
+    ) -> Result<bool> {
+        if self.bump_if(parser, TokenValue::LeftBracket).is_some() {
+            let size =
+                if let Some(Token { meta, .. }) = self.bump_if(parser, TokenValue::RightBracket) {
+                    span.subsume(meta);
+                    ArraySize::Dynamic
+                } else {
+                    let (value, constant_span) = self.parse_uint_constant(parser)?;
+                    let constant = parser.module.constants.fetch_or_append(
+                        crate::Constant {
+                            name: None,
+                            specialization: None,
+                            inner: crate::ConstantInner::Scalar {
+                                width: 4,
+                                value: crate::ScalarValue::Uint(value as u64),
+                            },
+                        },
+                        constant_span,
+                    );
+                    let end_span = self.expect(parser, TokenValue::RightBracket)?.meta;
+                    span.subsume(end_span);
+                    ArraySize::Constant(constant)
+                };
+
+            parser
+                .layouter
+                .update(&parser.module.types, &parser.module.constants)
+                .unwrap();
+            let stride = parser.layouter[*ty].to_stride();
+            *ty = parser.module.types.insert(
+                Type {
                     name: None,
-                    specialization: None,
-                    inner: crate::ConstantInner::Scalar {
-                        width: 4,
-                        value: crate::ScalarValue::Uint(value as u64),
+                    inner: TypeInner::Array {
+                        base: *ty,
+                        size,
+                        stride,
                     },
                 },
-                span,
+                *span,
             );
-            let end_meta = self.expect(parser, TokenValue::RightBracket)?.meta;
-            meta.subsume(end_meta);
-            Ok(Some((ArraySize::Constant(constant), meta)))
+
+            Ok(true)
```

Comment:
I don't understand why we are mutating `ty` and returning `Ok(bool)` if we could instead return something like `Result<Option<Handle<Type>>>`. Do you think this would be a feasible thing to change here?

Reply:
We definitely could but I don't what the benefit would be since we would be mutating the type anyways.

Reply:
Where would we be mutating the type? This would just return a new type.
Or do you mean that the caller would likely reassign the type anyway? That's OK, I think.

Reply:
Yes the caller always reassigns the type

Reply:
I think that would still be a better internal API.